File: c/005/005.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "unsigned_long_long_array.h"

void primeFactors(unsigned long long n, UnsignedLongLongArray *result);

int main(int argc, char *argv[]) {
  UnsignedLongLongArray primeFactorLists[20];

  UnsignedLongLongArray factors;

  UnsignedLongLongArrayInitialize(&factors);

  unsigned long long i;

  for (i = 1; i <= 20; i++) {
    UnsignedLongLongArrayInitialize(&primeFactorLists[i - 1]);

    primeFactors(i, &primeFactorLists[i - 1]);

    UnsignedLongLongArrayOverlap(&factors, &primeFactorLists[i - 1]);

    UnsignedLongLongArrayDeinitialize(&primeFactorLists[i - 1]);
  }

  unsigned long long product = 1;

  for (i = 0; i < factors.length; i++) {
    product *= factors.values[i];
  }

  printf("%llu\n", product);

  UnsignedLongLongArrayDeinitialize(&factors);

  return 0;
}

void primeFactors(unsigned long long n, UnsignedLongLongArray *result) {
  bool found = false;

  unsigned long long v = 2;

  for (unsigned long long i = 2; i < n - 1; i++) {
    if (n % i == 0) {
      found = true;

      v = i;

      break;
    }
  }

  if (found) {
    UnsignedLongLongArrayPush(result, v);

    primeFactors(n / v, result);
  } else {
    UnsignedLongLongArrayPush(result, n);
  }
}

File: csharp/005/005.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Euler {
    static public void Main() {
        List<List<ulong>> primeFactorLists = new List<List<ulong>>();

        for (ulong x = 1; x <= 20; x++) {
            primeFactorLists.Add(PrimeFactors(x));
        }

        List<ulong> factors = new List<ulong>();

        for (int i = 0; i < 20; i++) {
            factors = Overlap(factors, primeFactorLists[i]);
        }

        Console.WriteLine("{0}", factors.Aggregate((product, factor) => product * factor));
    }

    static public List<ulong> Overlap(List<ulong> a, List<ulong> b) {
        for (int i = 0; i < b.Count; i++) {
            a.Remove(b[i]);
        }

        a.AddRange(b);

        return a;
    }

    static public List<ulong> PrimeFactors(ulong n) {
        bool found = false;

        ulong v = 2;

        for (ulong i = 2; i < n - 1; i++) {
            if (n % i == 0) {
                found = true;

                v = i;

                break;
            }
        }

        List<ulong> result = new List<ulong>();

        if (found) {
            result.Add(v);

            result = result.Concat(PrimeFactors(n / v)).ToList();
        } else {
            result.Add(n);
        }

        return result;
    }
}
